K8SSAND-1088 ⁃ Add ability to disable user creation

<!--

Feel free to ask questions on Discord at https://discord.com/invite/y4s64xC9 or in the forum at https://forum.k8ssandra.io/. 
-->

**What is missing?**
cass-operator creates users, namely a default superuser, during the reconciliation process. In a multi-DC scenario where we are adding a DC to an existing cluster, we need to be to tell cass-operator to skip user creation. 

**Why do we need it?**
This is needed in k8ssandra-operator for https://github.com/k8ssandra/k8ssandra-operator/issues/21. We cannot add a DC to a C* 4.0 cluster without this change. 

In C* 4.0, you cannot update a keyspace's replication with a DC that does not yet exist. If I have an existing single-dc cluster, the replication of `system_auth` might look like this:

```
CREATE KEYSPACE system_auth WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '1'}  AND durable_writes = true;
```

If I add a second DC (i.e., a second CassandraDatacenter), the CassandraDatacenter will never become ready. Reconciliation will fail when trying to create the default superuser. The errors in Cassandra log look like this:

```
INFO  [epollEventLoopGroup-7-3] 2021-12-07 14:10:14,461 RpcMethod4x.java:142 - Failed to execute method NodeOps.createRole
java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at com.datastax.mgmtapi.rpc.RpcMethod4x.execute(RpcMethod4x.java:131)
	at com.datastax.mgmtapi.rpc.RpcMethod4x.execute(RpcMethod4x.java:36)
	at com.datastax.mgmtapi.interceptors.QueryHandlerInterceptor.lambda$handle$0(QueryHandlerInterceptor.java:130)
	at com.datastax.mgmtapi.shims.CassandraAPI.handleRpcResult(CassandraAPI.java:83)
	at com.datastax.mgmtapi.interceptors.QueryHandlerInterceptor.handle(QueryHandlerInterceptor.java:130)
	at com.datastax.mgmtapi.interceptors.QueryHandlerInterceptor.intercept(QueryHandlerInterceptor.java:93)
	at org.apache.cassandra.cql3.QueryProcessor.process(QueryProcessor.java)
	at org.apache.cassandra.transport.messages.QueryMessage.execute(QueryMessage.java:108)
	at org.apache.cassandra.transport.Message$Request.execute(Message.java:242)
	at org.apache.cassandra.transport.UnixSocketServer4x$UnixSockMessage.channelRead0(UnixSocketServer4x.java:105)
	at org.apache.cassandra.transport.UnixSocketServer4x$UnixSockMessage.channelRead0(UnixSocketServer4x.java:85)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at io.netty.channel.epoll.EpollDomainSocketChannel$EpollDomainUnsafe.epollInReady(EpollDomainSocketChannel.java:138)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.apache.cassandra.exceptions.UnavailableException: Cannot achieve consistency level LOCAL_ONE
```

The solution is pretty straightforward. Tell cass-operator to altogether skip creating users for the new CassandraDatacenter. I tested a POC that does this via an annotation. k8ssandra-operator adds that annotation to the new CassandraDatacenter.

**Environment**

* Cass Operator version:
1.9.0



┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1088) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1088
┆priority: Medium

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,7 +10,8 @@ Changelog for Cass Operator, new PRs should update the `main / unreleased` secti
 ```
 
 ## unreleased
-* [ENHANCEMENT] #235 Adding AdditionalLabels to add on all resources managed by the operator
+* [ENHANCEMENT] [#235](https://github.com/k8ssandra/cass-operator/issues/235) Adding AdditionalLabels to add on all resources managed by the operator
+* [ENHANCEMENT] [#244](https://github.com/k8ssandra/cass-operator/issues/244) Add ability to skip Casandra user creation
 
 ## v.1.9.0
 * [CHANGE] #202 Support fetching FeatureSet from management-api if available. Return RequestError with StatusCode when endpoint has bad status.

--- a/apis/cassandra/v1beta1/cassandradatacenter_types.go
+++ b/apis/cassandra/v1beta1/cassandradatacenter_types.go
@@ -44,6 +44,11 @@ const (
 	// ConfigHashAnnotation is the operator's annotation for the hash of the ConfigSecret
 	ConfigHashAnnotation = "cassandra.datastax.com/config-hash"
 
+	// SkipUserCreationAnnotation tells the operator to skip creating any Cassandra users
+	// including the default superuser. This is for multi-dc deployments when adding a
+	// DC to an existing cluster where the superuser has already been created.
+	SkipUserCreationAnnotation = "cassandra.datastax.com/skip-user-creation"
+
 	// CassNodeState
 	CassNodeState = "cassandra.datastax.com/node-state"
 

--- a/pkg/reconciliation/reconcile_racks.go
+++ b/pkg/reconciliation/reconcile_racks.go
@@ -893,6 +893,11 @@ func (rc *ReconciliationContext) UpdateSecretWatches() error {
 func (rc *ReconciliationContext) CreateUsers() result.ReconcileResult {
 	dc := rc.Datacenter
 
+	if val, found := dc.Annotations[api.SkipUserCreationAnnotation]; found && val == "true" {
+		rc.ReqLogger.Info(api.SkipUserCreationAnnotation + " is set, skipping CreateUser")
+		return result.Continue()
+	}
+
 	if dc.Spec.Stopped {
 		rc.ReqLogger.Info("cluster is stopped, skipping CreateUser")
 		return result.Continue()

